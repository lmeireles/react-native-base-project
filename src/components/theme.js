const colors = {
  white: "#EEE",
  light: "#B2B9C3",
  dark: "#26374B",
  primary: "#009A9A",
  secondary: "#FF2D66",
  danger: "#ff3f44"
};

const forms = {
  labelColor: colors.light,
  errorColor: colors.danger
};

const navigation = {
  headerBackground: colors.primary,
  tabBackground: colors.secondary,
  color: "#EEE",
  inactiveColor: "#EEEEEE88",
  iconSize: 28
};

const theme = Object.freeze({
  colors,
  forms,
  navigation,
  padding: 14,
  fontSize: 16
});

/*
  This function can be used to extract colors from components props
  Ex.: <MyComp primary />
  Inside "MyComp" we are able to call getColorFromProps(props) witch will return our theme`s primary color.
 */
export const getColorFromProps = (props) => {
  let finalColor = null;
  Object.keys(theme.colors).map(color => {
    if (color in props) {
      finalColor = theme.colors[color];
    }
  });

  return finalColor;
};

export default theme;
