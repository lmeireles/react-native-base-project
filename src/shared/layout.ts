import { FlexAlignType, FlexStyle } from 'react-native'
import { DefaultTheme } from 'styled-components'

type Spacing = number | 'view'
type ComputedSpacing = number | undefined

export interface LayoutProps {
  // Margin
  m?: Spacing
  mt?: Spacing
  mb?: Spacing
  ml?: Spacing
  mr?: Spacing
  mh?: Spacing
  mv?: Spacing

  // false = flex: undefined, true = flex: 1
  alignSelf?: FlexStyle['alignSelf']

  // Padding
  p?: Spacing
  pt?: Spacing
  pb?: Spacing
  pl?: Spacing
  pr?: Spacing
  ph?: Spacing
  pv?: Spacing
}

const computeSpacing = (
  v: Spacing | null | undefined,
  theme: DefaultTheme
): ComputedSpacing => {
  if (typeof v === 'number') {
    return theme.spacing(v)
  }

  if (v === 'view') {
    return theme.viewPadding
  }

  return undefined
}

export type ComputedLayoutStyles = {
  alignSelf: FlexAlignType | 'auto' | undefined

  margin: ComputedSpacing
  marginTop: ComputedSpacing
  marginBottom: ComputedSpacing
  marginLeft: ComputedSpacing
  marginRight: ComputedSpacing
  marginHorizontal: ComputedSpacing
  marginVertical: ComputedSpacing

  padding: ComputedSpacing
  paddingTop: ComputedSpacing
  paddingBottom: ComputedSpacing
  paddingLeft: ComputedSpacing
  paddingRight: ComputedSpacing
  paddingHorizontal: ComputedSpacing
  paddingVertical: ComputedSpacing
}

export const computeLayoutStyles = (
  inp: LayoutProps,
  theme: DefaultTheme
): ComputedLayoutStyles => ({
  alignSelf: inp.alignSelf,

  margin: computeSpacing(inp.m, theme),
  marginTop: computeSpacing(inp.mt, theme),
  marginBottom: computeSpacing(inp.mb, theme),
  marginLeft: computeSpacing(inp.ml, theme),
  marginRight: computeSpacing(inp.mr, theme),
  marginHorizontal: computeSpacing(inp.mh, theme),
  marginVertical: computeSpacing(inp.mv, theme),

  padding: computeSpacing(inp.p, theme),
  paddingTop: computeSpacing(inp.pt, theme),
  paddingBottom: computeSpacing(inp.pb, theme),
  paddingLeft: computeSpacing(inp.pl, theme),
  paddingRight: computeSpacing(inp.pr, theme),
  paddingHorizontal: computeSpacing(inp.ph, theme),
  paddingVertical: computeSpacing(inp.pv, theme),
})
