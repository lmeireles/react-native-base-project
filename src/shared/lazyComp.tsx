import * as React from 'react'
interface LazyCompState {
  Comp?: React.ComponentType
  crashed: boolean
}

export function lazyComp(loader: () => Promise<React.ComponentType<any>>): React.ReactNode {
  return class extends React.Component<any, LazyCompState> {
    state = {
      Comp: undefined,
      crashed: false,
    }
    mounted = false
    constructor(props: any) {
      super(props)
      // eslint-disable-next-line @typescript-eslint/no-this-alias
      const t = this
      loader().then(Comp => {
        if (!t.mounted) {
          return
        }
        if (!this.mounted) {
          return
        }
        t.setState({ Comp: (Comp as never) as React.ComponentType })
      })
    }
    componentDidMount() {
      this.mounted = true
    }
    componentWillUnmount() {
      this.mounted = false
    }
    componentDidCatch(error: Error, stack: React.ErrorInfo) {
      console.log('CATCH', error)
      console.log(stack)
      if (!this.mounted) {
        return
      }
      this.setState({
        crashed: true,
      })
    }

    render() {
      const { Comp, crashed } = this.state as LazyCompState
      if (Comp == null) {
        return null
      }
      if (crashed) {
        return null
      }
      return <Comp {...this.props} />
    }
  }
}
