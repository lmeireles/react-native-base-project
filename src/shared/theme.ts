import { Dimensions } from 'react-native'
import { DefaultTheme } from 'styled-components'
import * as styledComponents from 'styled-components/native'

export const {
  default: styled,
} = styledComponents as styledComponents.ReactNativeThemedStyledComponentsModule<DefaultTheme>

export const SMALL_DEVICE = Dimensions.get('window').width <= 350
export const MEDIUM_DEVICE = Dimensions.get('window').height <= 684
const UNIT = SMALL_DEVICE ? 4 : MEDIUM_DEVICE ? 6 : 8
const spacing = (factor: number): number => UNIT * factor

export const loadTheme = (externalTheme?: DefaultTheme): DefaultTheme => {
  return {
    spacing,
    viewPadding: SMALL_DEVICE ? spacing(2) : spacing(3),
    viewBackground: 'white',
    components: {
      textField: {
        height: 56,
        borderRadius: 12,
        background: 'gray07',
        labelColor: 'gray01',
        textColor: 'gray04',
      },
      ...(externalTheme?.components || {})
    },
    colors: {
      primary: '#240A36', // #364FD4
      secondary: '#e0e3fd',
      gray01: '#1E2126',
      gray02: '#5C6574',
      gray03: '#828B9A',
      gray04: '#B9C1CC',
      gray05: '#CAD1D9',
      gray06: '#E4E9ED',
      gray07: '#F0F3F8',
      white: '#FFFFFF',
      error: '#F13F11',
      success: '#00CC96',
      info: '#14C4E6',
      alert: '#FDC34E',
      ...(externalTheme?.colors || {})
    },
    text: {
      title: {
        family: 'Poppins-Bold',
        size: 28,
        spacing: 0.2,
      },
      smallTitle: {
        family: 'Poppins-Bold',
        size: 14,
      },
      subtitle: {
        family: 'Poppins-Regular',
        size: 16,
        spacing: 0.2,
      },
      body: {
        family: 'Poppins-Regular',
        size: 16,
        spacing: 0.1,
      },
      label: {
        family: 'Poppins-SemiBold',
        size: 16,
      },
      smallLabel: {
        family: 'Poppins-Regular',
        size: 12,
      },
      ...(externalTheme?.text || {})
    },
    buttons: {
      primary: {
        background: 'primary',
      },
      disabled: {
        background: 'gray05',
      },
      ...(externalTheme?.buttons || {})
    },
  }
}
