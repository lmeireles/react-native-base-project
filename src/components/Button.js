import React from 'react';
import {ActivityIndicator} from 'react-native';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import SimpleText from './SimpleText';
import theme, {getColorFromProps} from "./theme";

const ButtonContainer = styled.TouchableOpacity`
  width: 100%;
  padding: ${ p => p.padding}px;
  background: ${ p => p.backgroundColor};
  justify-content: center;
  align-items: center;
  border-radius: ${p => p.round ? 7 : 2}px;
`;

const Button = (props) => {
  const {loading, title, color, ...rest} = props;

  rest.backgroundColor = getColorFromProps(props) || rest.backgroundColor;

  return (
    <ButtonContainer {...rest}>
      { loading ?
        <ActivityIndicator color={color} />
        :
        <SimpleText color={color} bold>{title}</SimpleText>
      }
    </ButtonContainer>
  )
};

Button.propTypes = {
  backgroundColor: PropTypes.string,
  color: PropTypes.string,
  padding: PropTypes.number,
  round: PropTypes.bool,
  loading: PropTypes.bool,
  title: PropTypes.string
};

Button.defaultProps = {
  backgroundColor: theme.colors.light,
  color: '#EEE',
  padding: theme.padding,
  round: false,
  loading: false,
};

export default Button;
