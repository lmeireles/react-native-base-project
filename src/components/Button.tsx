import React, { FC } from 'react'
import { ButtonVariants, useTheme } from 'styled-components'
import {
  StyleProp,
  TouchableOpacity,
  TouchableOpacityProps,
  ViewStyle,
} from 'react-native'
import { Text } from 'components/Text'
import { computeLayoutStyles, LayoutProps } from 'shared/layout'

interface ButtonProps extends TouchableOpacityProps {
  variant?: keyof ButtonVariants
}
const Button: FC<ButtonProps & LayoutProps> = ({
  variant = 'primary',
  disabled,
  children,
  style,
  ...props
}) => {
  const theme = useTheme()
  const btnStyle = theme.buttons[disabled ? 'disabled' : variant]
  const hasBorder = !!btnStyle.borderWidth && !!btnStyle.borderColor

  const styles = [
    computeLayoutStyles(props, theme),
    {
      alignItems: 'center',
      justifyContent: 'center',
      width: '100%',
      height: 54,
      borderRadius: 12,
      backgroundColor: theme.colors[hasBorder ? 'white' : btnStyle.background],
      borderWidth: hasBorder ? btnStyle.borderWidth : 0,
      borderColor: theme.colors[btnStyle.borderColor || 'primary'],
    } as StyleProp<ViewStyle>,
    style,
  ]

  return (
    <TouchableOpacity {...props} disabled={disabled} style={styles}>
      <Text
        variant="label"
        color={hasBorder ? btnStyle.borderColor || 'primary' : 'white'}
      >
        {children}
      </Text>
    </TouchableOpacity>
  )
}

export default Button
