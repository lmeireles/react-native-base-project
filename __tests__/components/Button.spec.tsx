import React from 'react'
import { render, snapshot } from '../test-utils'
import { loadTheme } from 'shared/theme'
import 'jest-styled-components/native'
import { ButtonVariants } from 'styled-components'
import Button from 'components/Button'
import { fireEvent } from '@testing-library/react-native'

describe('Button snapshots', () => {
  const theme = loadTheme()

  it('default', () => {
    snapshot(render(<Button>Test</Button>))
  })

  it('variants', () => {
    Object.keys(theme.buttons).map(key => {
      snapshot(
        render(<Button variant={key as keyof ButtonVariants}>Test</Button>)
      )
    })
  })

  it('dimensions', () => {
    snapshot(
      render(
        <Button m={3} pt={5} ph={6}>
          Test
        </Button>
      )
    )
  })
})

describe('Button', () => {
  test('onPress callback', () => {
    const cb = jest.fn()
    const { getByText } = render(<Button onPress={cb}>Test</Button>)

    fireEvent.press(getByText('Test'))
    expect(cb).toBeCalled()
  })
})
