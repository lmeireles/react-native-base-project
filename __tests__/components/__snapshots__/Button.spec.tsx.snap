// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshots default 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": undefined,
      "backgroundColor": "#240A36",
      "borderColor": "#240A36",
      "borderRadius": 12,
      "borderWidth": 0,
      "height": 54,
      "justifyContent": "center",
      "margin": undefined,
      "marginBottom": undefined,
      "marginHorizontal": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "marginVertical": undefined,
      "opacity": 1,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
      "width": "100%",
    }
  }
>
  <Text
    maxFontSizeMultiplier={1.1}
    style={
      Array [
        Object {
          "alignSelf": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginHorizontal": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "marginVertical": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "paddingVertical": undefined,
        },
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Poppins-SemiBold",
          "fontSize": 16,
          "letterSpacing": undefined,
          "textAlign": "auto",
        },
        undefined,
      ]
    }
  >
    Test
  </Text>
</View>
`;

exports[`Button snapshots dimensions 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": undefined,
      "backgroundColor": "#240A36",
      "borderColor": "#240A36",
      "borderRadius": 12,
      "borderWidth": 0,
      "height": 54,
      "justifyContent": "center",
      "margin": 24,
      "marginBottom": undefined,
      "marginHorizontal": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "marginVertical": undefined,
      "opacity": 1,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": 48,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": 40,
      "paddingVertical": undefined,
      "width": "100%",
    }
  }
>
  <Text
    maxFontSizeMultiplier={1.1}
    style={
      Array [
        Object {
          "alignSelf": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginHorizontal": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "marginVertical": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "paddingVertical": undefined,
        },
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Poppins-SemiBold",
          "fontSize": 16,
          "letterSpacing": undefined,
          "textAlign": "auto",
        },
        undefined,
      ]
    }
  >
    Test
  </Text>
</View>
`;

exports[`Button snapshots variants 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": undefined,
      "backgroundColor": "#240A36",
      "borderColor": "#240A36",
      "borderRadius": 12,
      "borderWidth": 0,
      "height": 54,
      "justifyContent": "center",
      "margin": undefined,
      "marginBottom": undefined,
      "marginHorizontal": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "marginVertical": undefined,
      "opacity": 1,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
      "width": "100%",
    }
  }
>
  <Text
    maxFontSizeMultiplier={1.1}
    style={
      Array [
        Object {
          "alignSelf": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginHorizontal": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "marginVertical": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "paddingVertical": undefined,
        },
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Poppins-SemiBold",
          "fontSize": 16,
          "letterSpacing": undefined,
          "textAlign": "auto",
        },
        undefined,
      ]
    }
  >
    Test
  </Text>
</View>
`;

exports[`Button snapshots variants 2`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": undefined,
      "backgroundColor": "#CAD1D9",
      "borderColor": "#240A36",
      "borderRadius": 12,
      "borderWidth": 0,
      "height": 54,
      "justifyContent": "center",
      "margin": undefined,
      "marginBottom": undefined,
      "marginHorizontal": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "marginVertical": undefined,
      "opacity": 1,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
      "width": "100%",
    }
  }
>
  <Text
    maxFontSizeMultiplier={1.1}
    style={
      Array [
        Object {
          "alignSelf": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginHorizontal": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "marginVertical": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "paddingVertical": undefined,
        },
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Poppins-SemiBold",
          "fontSize": 16,
          "letterSpacing": undefined,
          "textAlign": "auto",
        },
        undefined,
      ]
    }
  >
    Test
  </Text>
</View>
`;
